src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -650,6 +621,8 @@ internal bool TryRead(out ReadResult result)
                 {
                     ThrowHelper.ThrowInvalidOperationException_AlreadyReading();
                 }
+
+                _operationState.BeginReadTentative();
```

Comment:
This is new. If TryRead returns false, we want to avoid throwing in EndRead in case you do:

```
TryRead()
AdvanceTo()
```

Reply:
Hm, shouldn't it be conditional then? In the case when TryRead returns true you should always call Advance. 
+ Add a test.